theano/tensor/nnet/abstract_conv.py

```diff
@@ -2314,29 +2554,62 @@ def perform(self, node, inp, out_):
-            mshp1 = mat.shape[1] * self.num_groups
+            mshp1 = mat.shape[-self.convdim - 1] * self.num_groups
             mat = mat.reshape((self.num_groups, mshp0) + mat.shape[1:])
-            mat = mat.transpose((1, 0, 2) + tuple(range(3, 3 + self.convdim)))
-            mat = mat.reshape((mshp0, mshp1) + mat.shape[-self.convdim:])
+            if self.unshared:
+                # for 2D -> (1, 2, 3, 0, 4, 5, 6)
+                mat = mat.transpose(tuple(range(1, 2 + self.convdim)) + (0,) +
+                                    tuple(range(2 + self.convdim, mat.ndim)))
+                mat = mat.reshape((mshp0,) + mat.shape[1:1 + self.convdim] + (mshp1,) + mat.shape[-self.convdim:])
+            else:
+                mat = mat.transpose((1, 0, 2) + tuple(range(3, 3 + self.convdim)))
+                mat = mat.reshape((mshp0, mshp1) + mat.shape[-self.convdim:])
             return mat
+
         kern = correct_for_groups(kern)
-        kern = kern.transpose(axes_order)
 
-        if self.filter_flip:
-            topgrad = topgrad[flip_filters]
-        img = self.conv(topgrad, kern, mode="full", dilation=self.filter_dilation, num_groups=self.num_groups)
-        if self.filter_flip:
-            img = img[flip_filters]
+        if self.unshared:
+            # from (nFilters, out_rows, out_cols, nChannels, kH, kW)
+            # to (nChannels, nFilters, out_rows, out_cols, kH, kW)
+            axes_order = (1 + self.convdim, 0,) + tuple(range(1, 1 + self.convdim)) + \
+                tuple(range(2 + self.convdim, kern.ndim))
+            kern = kern.transpose(axes_order)
+            if not self.filter_flip:
+                kern = kern[(slice(None),) * (kern.ndim - self.convdim) +
+                            (slice(None, None, -1),) * self.convdim]
+            img = self.conv(topgrad, kern, mode="full", dilation=self.filter_dilation,
+                            num_groups=self.num_groups, unshared=True, direction="backprop inputs")
+        else:
+            axes_order = (1, 0) + tuple(range(2, 2 + self.convdim))
+            kern = kern.transpose(axes_order)
+            flip_filters = ((slice(None), slice(None)) +
+                            (slice(None, None, -1),) * self.convdim)
+            if self.filter_flip:
+                topgrad = topgrad[flip_filters]
+            img = self.conv(topgrad, kern, mode="full", dilation=self.filter_dilation,
+                            num_groups=self.num_groups)
+            if self.filter_flip:
+                img = img[flip_filters]
+
         if any(p > 0 for p in pad):
             img = img[(slice(None), slice(None)) +
                       tuple(slice(pad[i], img.shape[i + 2] - pad[i])
                             for i in range(self.convdim))]
+        if mode == 'causal':
+            left_pad = self.filter_dilation[0] * (kern.shape[-self.convdim] - 1)
+            img = img[(slice(None), slice(None),
+                      slice(left_pad, left_pad + img.shape[2]))]
```

Comment:
For example, these lines could be merged with the normal padding logic if you split `pad` into `pad_left` and `pad_right`, which would be equal for normal padding and different for causal. (My argument for this would be that that it doesn't make it harder to implement causal convolution now, but would make it easier to add support for asymmetric padding at some point.)